lib/fieldy/writer.rb

```diff
@@ -4,38 +4,46 @@ module Writer
 
     def self.included(base)
       base.extend(WriterMethods)
+      base.send :include, InstanceMethods
+    end
+
+    module InstanceMethods
+
+      def to_s
+        data = self.class.fields
+                         .map        { |x| x[:key] }
+                         .select     { |x| x }
+                         .reduce({}) { |x, k| x.merge! k => self.send(k) }
+        self.class.write data
+      end
+
     end
-      
 
     module WriterMethods
-      
+
       def write(hash)
-        res = []
-        pack_str = ""
-        @fields.each do |h|
-          h.each do |k,v|
-            hash.each do |x,y|
-              unless k == :null
-                res << y
-              end
-            end
-          pack_str << "#{v[:type]}#{v[:length]}"
-          end
-        end
-        res.pack(pack_str)
+        values  = fields.map { |x| x[:key] }.map { |x| hash[x] || '' }
+        pack    = fields.map { |x| "#{x[:type]}#{x[:length]}" }.join ''
+
+        values.pack pack
       end
-      
-      def field(sym, length, type = "A")
+
+      def fields
         @fields ||= []
-        @fields << { sym => { :length => length, :type => type }}
       end
-      
+
+      def field(sym, length, type = "A")
+        starts_at = fields.reduce(0) { |t, i| t + i[:length] }
+        fields << { key: sym, length: length, starts_at: starts_at, type: type }
+        self.send(:attr_accessor, sym) if sym
+      end
+
       def skip(length)
-        self.field(:null, length, "x" )
+        self.field(nil, length, "A" )
```

Comment:
I had a big issue with the spaces... I wanted the "skip" to just fill with spaces, but it filled it with those nil characters or something like that.  I changed the approach to use ```" "``` instead.